<a href="https://colab.research.google.com/github/GeminiNetSailor/Data-science-and-analytics/blob/main/7.2%20Actividad%207.%20PCA/Actividad7PCA_Equipo04.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**MAESTRÍA EN INTELIGENCIA ARTIFICIAL APLICADA**

**Curso: TC4029 - Ciencia y analítica de datos**

Tecnológico de Monterrey

Prof Grettel Barceló Alonso

**Semana 7**
Análisis de Componentes Principales (PCA)

---

*   NOMBRE: David A. Serrano Garcia
*   MATRÍCULA: A01795935
*   NOMBRE: Said Chávez Cervantes
*   MATRÍCULA: A01114101
*   NOMBRE: Erik Abel Cruz Yekon
*   MATRÍCULA: A01795929

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
# Define enviroments constant names
DEFAULT = 'DEFAULT'
VARIANT = 'VARIANT'

# Define the environment settings in a dictionary
ENVIRONMENTS_CONFIG = {
    DEFAULT: {
        'base_path': '/content/drive/MyDrive/Colab Notebooks/MNA/TC4029 - Ciencia y analítica de datos/Semana 7/Actividad7_PCA'
    },
    VARIANT: {
        'base_path': '/content/drive/MyDrive/Colab Notebooks/Maestría MNA/Ciencia y Analítica de datos'
    }
}


In [3]:
# Descomenta la configuración que vas a usar.
environment = DEFAULT
#environment = VARIANT

In [4]:
base_path = ENVIRONMENTS_CONFIG[environment]['base_path']

In [5]:
import os
DIR = base_path
os.chdir(DIR)

In [6]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats

In [7]:
cereals_df = pd.read_csv('Cereals.csv')
cereals_df

name mfr type  calories  protein  fat  sodium  fiber  \
0                   100%_Bran   N    C        70        4    1     130   10.0   
1           100%_Natural_Bran   Q    C       120        3    5      15    2.0   
2                    All-Bran   K    C        70        4    1     260    9.0   
3   All-Bran_with_Extra_Fiber   K    C        50        4    0     140   14.0   
4              Almond_Delight   R    C       110        2    2     200    1.0   
..                        ...  ..  ...       ...      ...  ...     ...    ...   
72                    Triples   G    C       110        2    1     250    0.0   
73                       Trix   G    C       110        1    1     140    0.0   
74                 Wheat_Chex   R    C       100        3    1     230    3.0   
75                   Wheaties   G    C       100        3    1     200    3.0   
76        Wheaties_Honey_Gold   G    C       110        2    1     200    1.0   

    carbo  sugars  potass  vitamins  shelf  weight  cups     rating  
0     5.0     6.0   280.0        25      3     1.0  0.33  68.402973  
1     8.0     8.0   135.0         0      3     1.0  1.00  33.983679  
2     7.0     5.0   320.0        25      3     1.0  0.33  59.425505  
3     8.0     0.0   330.0        25      3     1.0  0.50  93.704912  
4    14.0     8.0     NaN        25      3     1.0  0.75  34.384843  
..    ...     ...     ...       ...    ...     ...   ...        ...  
72   21.0     3.0    60.0        25      3     1.0  0.75  39.106174  
73   13.0    12.0    25.0        25      2     1.0  1.00  27.753301  
74   17.0     3.0   115.0        25      1     1.0  0.67  49.787445  
75   17.0     3.0   110.0        25      1     1.0  1.00  51.592193  
76   16.0     8.0    60.0        25      1     1.0  0.75  36.187559  

[77 rows x 16 columns]

# **Parte 1**. EDA y preprocesamiento

1. Obtén estadísticas descriptivas para todas las variables del dataframe.

In [8]:
dfn = cereals_df.select_dtypes(include=np.number).columns.tolist()
dfc = cereals_df.select_dtypes(exclude=np.number).columns.tolist()

In [9]:
cereals_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77 entries, 0 to 76
Data columns (total 16 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   name      77 non-null     object 
 1   mfr       77 non-null     object 
 2   type      77 non-null     object 
 3   calories  77 non-null     int64  
 4   protein   77 non-null     int64  
 5   fat       77 non-null     int64  
 6   sodium    77 non-null     int64  
 7   fiber     77 non-null     float64
 8   carbo     76 non-null     float64
 9   sugars    76 non-null     float64
 10  potass    75 non-null     float64
 11  vitamins  77 non-null     int64  
 12  shelf     77 non-null     int64  
 13  weight    77 non-null     float64
 14  cups      77 non-null     float64
 15  rating    77 non-null     float64
dtypes: float64(7), int64(6), object(3)
memory usage: 9.8+ KB


In [ ]:
cereals_df.describe(include = 'object').T

In [ ]:
cereals_df.describe().T

Genera algunos gráficos para familiarizarte con el conjunto de datos. Al menos deberás incluir los que te permitan responder los siguientes cuestionamientos:

**Nota.** Debes responder de manera explícita las preguntas, apoyándote de los resultados observados de cada gráfico.

2a) ¿Cuál es la frecuencia o conteo de cada categoría para las variables: `mfr` y `shelf` (por separado)?

**mfr:**

**Kelloggs -	23**

**General Mills -	22**

**Post -	9**

**Quaker Oats -	8**

**Ralston Purina -	8**

**Nabisco -	6**

**American Home Food Products -	1**

**Shelf:**

**1 -	20**

**2	- 21**

**3 -	36**

In [ ]:
cereals_df['mfr'].value_counts().to_frame()

In [ ]:
cereals_df['shelf'].value_counts().to_frame()

In [ ]:
for column in dfc:
    print(cereals_df[column].value_counts())
    print('-' * 50)

In [ ]:
plt.figure(figsize = (8, 5))
sns.countplot(x='mfr', data=cereals_df)

In [ ]:
plt.figure(figsize = (8, 5))
sns.countplot(x='shelf', data=cereals_df)

2b) Combina los resultados previos para observar, de cada fabricante, cuántos productos tiene en cada estante.

**Kelloggs -	4 en Shelf 1, 7 en Shelf 2 y 12 en Shelf 3.**

**General Mills -	6 en Shelf 1, 7 en Shelf 2 y 9 en Shelf 3.**

**Post -	2 en Shelf 1, 1 en Shelf 2 y 6 en Shelf 3.**

**Quaker Oats -	1 en Shelf 1, 3 en Shelf 2 y 4 en Shelf 3.**

**Ralston Purina -	4 en Shelf 1, 0 en Shelf 2 y 4 en Shelf 3.**

**Nabisco -	3 en Shelf 1, 2 en Shelf 2 y 1 en Shelf 3.**

**American Home Food Products -	0 en Shelf 1, 1 en Shelf 2 y 0 en Shelf 3.**

In [ ]:
plt.figure(figsize = (8, 6))
sns.countplot(x='mfr', data=cereals_df, hue='shelf')

 3a) ¿Cuál es el rango más frecuente de la variable `rating`?

**El rango más frecuente de la variable rating es entre 33.174094 y 50.828392, tal cómo se muestra en la gráfica.**

In [ ]:
Q1 = cereals_df['rating'].quantile(0.25)
Q3 = cereals_df['rating'].quantile(0.75)
print("El rango más frecuente de la variable rating es entre",Q1,"y",Q3)

In [ ]:
sns.boxplot(x=cereals_df['rating'])

3b) ¿Cuáles son los 10 cereales mejor evaluados y a qué fabricante corresponden?

**1.- All-Bran_with_Extra_Fiber -	Kellogs	(93.704912).**

**2.- Shredded_Wheat_'n'Bran -	Nabisco	(74.472949).**

**3.- Shredded_Wheat_spoon_size	- Nabisco	(72.801787).**

**4.- 100%_Bran	- Nabisco	(68.402973).**

**5.- Shredded_Wheat -	Nabisco	(68.235885).**

**6.- Cream_of_Wheat_(Quick) -	Nabisco	(64.533816).**

**7.- Puffed_Wheat -	Quaker Oats	(63.005645).**

**8.- Puffed_Rice	Quaker - Oats	(60.756112).**

**9.- Nutri-grain_Wheat	- Kellogs	(59.642837).**

**10.- All-Bran	Kellogs -	(59.425505).**

In [ ]:
cereals_df[['name','mfr','rating']].sort_values('rating', ascending=False)[0:10]

In [ ]:
plt.figure(figsize=(12, 6))
sns.barplot(x=cereals_df['name'], y=cereals_df['rating'])
plt.xticks(rotation=90)
plt.xlabel('Cereal')
plt.ylabel('Rating')
plt.title('Evaluación de cereales')
plt.show()

In [ ]:
cereals_df[['name','mfr','rating']]

4a) Compara la distribución de las calorías según el fabricante, por medio de boxplots.

**Se puede observar que el fabricante Nabisco tiene el menor rango de calorias registrado, mientras que llama la atención cómo el nivel de calorias de American Home Food Products y General Mills es bastante constante.**

In [ ]:
sns.boxplot(x=cereals_df['calories'],y=cereals_df['mfr'])

4b) ¿Cuáles son las dos variables que tienen mayor correlación con `rating`?

**Fibra (0.58) y Proteína (0.47).**

In [ ]:
plt.figure(figsize = (12, 10))
sns.heatmap(round(cereals_df.corr(numeric_only=True),2), annot = True)

5. Elimina todos los registros con algún dato faltante y reinicia el índice del dataframe para que quede con valores consecutivos.

In [ ]:
cereals_df2 = cereals_df.dropna().reset_index(drop=True)
cereals_df2

# **Parte 2**. Ingeniería de características

6. Considerando que `rating` es la variable de salida, almacénala en una variable `y` y separa los predictores **numéricos** en `X`. Escala los valores de `X` y aplica `PCA` para proyectar los datos en el nuevo espacio de vectores.

In [ ]:
#seprara rl objetivo
y=cereals_df2.loc[:,'rating']
#serpara los predictores
x=cereals_df2.iloc[:,3:-1]

In [ ]:
total_componentes=x.shape[1]

In [ ]:
#escalar los datos
from sklearn.preprocessing import StandardScaler
escalar=StandardScaler()
x_scaled=escalar.fit_transform(x)
x_scaled[0:5]

In [ ]:
#Aplicacion de PCA
from sklearn.decomposition import PCA
pca=PCA()
x_projected=pca.fit_transform(x_scaled)
x_projected=pd.DataFrame(x_projected)

In [ ]:
#Nuevo espacio de Vectores
print("Valores propios", list(np.round(pca.explained_variance_,decimals=8)))

7. Obtén la curva del porcentaje de varianza acumulada y determine el número mínimo de componentes principales que explique más del 90% de la varianza. Imprime la información de dichos componentes.

Para obtener mas del 90% de de la varianza debe de tener un minimo de 7 componentes

In [ ]:
print("Varianza acumulada", np.cumsum(pca.explained_variance_ratio_))

8. ¿Cuáles son las tres variables más importantes en el cálculo del primer componente?


In [ ]:
num_components=12
pc_df=pd.DataFrame(abs(pca.components_[:num_components]), columns=x.columns, index=['Principal componente {}'.format(i) for i in range(1, num_components+1)])
pc_df

9. ¿Todas las variables categóricas son predictores? Aplica una técnica de encoding a las que sí lo sean. Justifica la elección de tu técnica.

10. Conjunta, en un dataframe, las valores proyectados en los componentes seleccionados (mínimo), las transformaciones obtenidas de las variables categóricas que serán predictores y la variable de salida. Almacena el dataframe resultante en archivo.